packages/babel-plugin-transform-react-jsx-self/src/index.js

```diff
@@ -16,11 +16,78 @@ import { types as t } from "@babel/core";
 
 const TRACE_ID = "__self";
 
+/**
+ * Returns whether the given path contains a `super()` call.
+ */
+function containsSuperCall(sourcePath) {
+  // We don't want to go into inner classes as their super() calls aren't relevant for us.
+  if (sourcePath.isClassDeclaration() || sourcePath.isClassExpression()) {
+    return false;
+  }
+  let exists = false;
+  sourcePath.traverse({
+    // Again, we don't go into inner classes.
+    ClassDeclaration(path) {
+      path.skip();
+    },
+    ClassExpression(path) {
+      path.skip();
+    },
+    // Find a `Super` that is used in a call expression.
+    Super(path) {
+      if (path.parentPath.isCallExpression()) {
+        exists = true;
+        path.stop();
+      }
+    },
+  });
+  return exists;
+}
+
+/**
+ * Returns whether it is allowed to use `this` at given path.
+ */
+function canUseThis(path) {
+  const statement = path.getStatementParent();
+  // This specifically skips arrow functions as they do not rewrite `this`.
+  const parentMethodOrFunction = statement.findParent(path => {
```

Comment:
Consider traverse the `scope.parent`, it is faster because it skips nodes which do not create a scope.

```js
// derived from `Scope#getFunctionParent`
getThisParent(): Scope | null {
    let scope: Scope = this;
    do {
      if (scope.path.isFunctionParent() && !scope.path.isArrowFunction()) {
        return scope;
      }
    } while ((scope = scope.parent));
    return null;
  }
```

The `parentMethodOrFunction` is then `path.scope.getThisFunctionParent().path`.

Reply:
Done.